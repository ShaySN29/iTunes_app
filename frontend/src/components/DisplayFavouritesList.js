import React from 'react';
import { useState, useEffect } from 'react';
// Import Icon
import homeIcon from '../images/homeIcon.png'
// Import from bootstrap
import { Card, Button } from 'react-bootstrap';

function DisplayFavouritesList({ handleHomeButtonClick }) {
  // Setting state of the favourite items array to an empty array to store the favouriteList from session storage in
  const [favouriteItems, setFavouriteItems] = useState([]);

  // If the favouritesList is not in storage then it gets set in storage
  // If the favouritesList is in storage then it is retrieved and parsed and set in state
  function initialiseFavouritesList() {
    if (sessionStorage.getItem("favouritesList") === null) {
      sessionStorage.setItem("favouritesList", JSON.stringify(favouriteItems));
    } else {
      setFavouriteItems(JSON.parse(sessionStorage.getItem("favouritesList")));
    };
  };

  /* useEffect used to prevent an infinite loop.
    I used this link to help: https://www.w3schools.com/react/react_useeffect.asp
  */
  useEffect(() => {
    initialiseFavouritesList();
  }, []);

  //  Function below maps through the favouriteItems array and displays it in a Card
  //  Delete button calls the deleteFavouriteitem function
  function getListOfFavourites() {
    return (
      favouriteItems.map((result, index) => {
        return (
          <div className='individualCard' key={result.myKey}>
            <Card style={{ width: '18rem' }} border="success" className="displayCard mb-3 ms-2">
              <Card.Body>
                <Card.Title>{result.name}</Card.Title>
                <Card.Subtitle className="mb-2 text-muted">{result.artist}</Card.Subtitle>
                <Card.Img variant="top" src={result.artwork} />
                <Card.Link href={result.link}>More Details</Card.Link>
              </Card.Body>
              <Button variant="outline-success" onClick={() => deleteFavouriteItem(index)}>Remove</Button>
            </Card>
          </div>
        )
      }
    ));
  };

  // The function belows deletes an item from the favourites list
  function deleteFavouriteItem(itemIndex) {
    // Deleting the selected item from favourites using the index
    favouriteItems.splice(itemIndex, 1);

    // updating session storage
    sessionStorage.setItem("favouritesList", JSON.stringify(favouriteItems));

    // setting state of the updated list
    setFavouriteItems(favouriteItems);
    
    // Updating the list of the displayed items
    initialiseFavouritesList();

    alert("Your item has been deleted");
  };

  return (
    <div>
      <div className="iconDiv">
        <img src={homeIcon} height={40} className="mt-1 mb-3" />
      </div>
      <div>
        <Button variant="success" className="mt-2 mb-3" onClick={(e) => handleHomeButtonClick()}>Home</Button>
        <h3 className="favouritesHeading bg-success">My Favourites</h3>
      </div>
      <div>
        {/* The consitional operator below sets the condition that when the length of the favouriteItems array is greater than 0 then 
            the list of favourites will be displayed else the message telling the user to add favourites  is displayed */}
        {favouriteItems.length !== 0 ? getListOfFavourites() : <h3 className="emptyListOfFavouritesText"> You don't have anything in your favourites yet. <br /> You can search for items by clicking the home button above. <br /> Happy Searching!!! </h3>}
      </div>
    </div>
  );
};

export default DisplayFavouritesList;

/* ACKNOWLEDGMENTS
Home Icon: https://www.vecteezy.com/vector-art/2276894-vector-illustration-of-home-icon */
