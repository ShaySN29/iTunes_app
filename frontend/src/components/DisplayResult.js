import React, { useEffect } from 'react';
// Import from Bootstrap
import { Button, Card } from 'react-bootstrap';

    /* Functions below checks if item is in favourites by comparing the myKey property (unique property) 
        it returns true or false depending on if the item exist in the favourite list.

        The function is outside of the functional component so I could test the function.
        I used the below links:
        https://waresix.engineering/testing-your-react-component-with-react-testing-library-e40062abaf36
        https://medium.com/the-fours/writing-functions-outside-vs-inside-in-react-c9044ea31ee2#:~:text=Writing%20functions%20outside%20component%20makes,does%20not%20rely%20on%20props.
    */
        function checkIfFavouriteItemExists(favouriteItems, myKey) {
            return favouriteItems.some(element => {
                if (element.myKey === myKey) {
                    return true;
                } else {
                    return false;
                }
            });
        };

// The DisplayResult component displays each of the results from the search into a card
// Props are passed from the SearchBar component
function DisplayResult(props) {
    let favouriteItems = []; // Empty array to store the favourite items in
    
    // If the favouritesList is not in storage then it gets set in storage
    // If the favouritesList is in storage then it is retrieved and parsed
    function initialiseFavouritesList() {
        if (sessionStorage.getItem("favouritesList") === null) {
            sessionStorage.setItem("favouritesList", JSON.stringify(favouriteItems));
        } else {
            favouriteItems = JSON.parse(sessionStorage.getItem("favouritesList"));
        };
    };

    /* useEffect used to prevent an infinite loop.
    I used this link to help: https://www.w3schools.com/react/react_useeffect.asp
    */
    useEffect(() => {
        initialiseFavouritesList();
    }, []);
    


    /* The function below calls the checkIfFavouriteItemExists that is outside the functional
    If the item is in favourites - the user is alerted the item is already in favourites
    If the item is not in favourites then the addToFavourites function is called
    I used this link to assist: https://bobbyhadz.com/blog/javascript-check-if-array-contains-object 
    */
    function isItemInFavourites() {        
        const isFound = checkIfFavouriteItemExists(favouriteItems, props.myKey);

        // If is found is true then the user is alerted that the item is already in favourites else the item is added to favourites
        if (isFound) {
            alert("The item is already in your favourites!");
        } else {
            addToFavourites();
        };
    };

    // The function below adds the item to the favourites list if it does not already exist in the list
    function addToFavourites() {
        // Creating a new object that will be passed into our session storage.
        const newFavouriteItem = {
            id: props.id,
            myKey: props.myKey, 
            link: props.link,
            name: props.name,
            artist: props.artist,
            artwork: props.artwork,
        };

        // Using the array.push() method to add the new favourite item in the favourites array
        favouriteItems.push(newFavouriteItem); 

        // updating session storage
        sessionStorage.setItem("favouritesList", JSON.stringify(favouriteItems));
        
        // Alerting the user that the item has been added
        alert("Your item has been added to Favourites!");
    };

    return (
        <div className="individualCard" key={props.key}>
            <Card style={{ width: '18rem' }} border="success" className="displayCard mb-3 ms-2">
                <Card.Body>
                    <Card.Title>{props.name}</Card.Title>
                    <Card.Subtitle className="mb-2 text-muted">{props.artist}</Card.Subtitle>
                    <Card.Img variant="top" src={props.artwork} />
                    <Card.Link href={props.link}>More Details</Card.Link>
                </Card.Body>
                <Button variant="outline-success" onClick={() => isItemInFavourites()}>Add to Favourites</Button>
            </Card>
        </div>
    );
};

// export { DisplayResult, checkIfFavouriteItemExists };
export { DisplayResult, checkIfFavouriteItemExists };