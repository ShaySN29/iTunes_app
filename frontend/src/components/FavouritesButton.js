import React from 'react';
// Imports from React Bootstrap
import { Button } from 'react-bootstrap';
// Importing the favourite icon
import favouriteIcon from '../images/favouriteIcon.png'

/* The FavouritesButton functional component displays the favourites button
  When the favourites button is clicked the state of the display favouritesList is set to true and the favourites list is shown */
function FavouritesButton({handleFavouriteButtonClick}) {
  return (
    <div className="favouriteBarDiv">
      <div className="iconDiv">
        <img src={favouriteIcon} height={40} className="mt-1 mb-3"/>
      </div>
      <div>
        <Button variant="success" className="mt-2 mb-3" onClick={e => handleFavouriteButtonClick()}>Favourites</Button>
      </div>
    </div>
  )
};

export default FavouritesButton;

/* ACKNOWLEDGEMENTS
Favourites icon used from: https://www.vecteezy.com/vector-art/379885-favorite-vector-icon */