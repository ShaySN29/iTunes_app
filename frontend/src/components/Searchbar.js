import React, { Component } from 'react';
// Imports from React Bootstrap
import { Navbar } from 'react-bootstrap';
import { Form, Button, Container } from 'react-bootstrap';
// Import Component
import {DisplayResult} from './DisplayResult';
require('isomorphic-fetch'); // Used to fetch API from backend

/* The SearchBar Class component contains the search bar and media option dropdown. It makes the call to the backend server to fetch the 
    input of the user. 
    Props are sent to the Display Result component from mapping through the items returned from the search 
*/
class Searchbar extends Component {
    constructor(props) {
        super(props);
        this.handleSearchBarChange = this.handleSearchBarChange.bind(this);
        this.handleMediaSelection = this.handleMediaSelection.bind(this);
        this.handleSearchBarSubmit = this.handleSearchBarSubmit.bind(this);
        this.displaySearchResults = this.displaySearchResults.bind(this);
        this.state = {
            searchBar: "", // The state of the search bar is an empty string
            mediaOption: "all", // Initial state of the Media selected
            itemsReturnedFromSearch: [], // Empty array to store the items returned from the search
            error: null
        }
    };

    /* When the user inputs a request into the searchBar, the state of the search bar is changed to the value of tuser input */
    handleSearchBarChange(event) {
        this.setState({ searchBar: event.target.value })
    };

    /* When the user selects a media type from the options the state is changed to the selection, if no option is selected the default is all */
    handleMediaSelection(event) {
        this.setState({mediaOption: event.target.value})
    };

    /* When the search button is clicked:
        If the search bar is empty, the user is alerted to enter an item to search for.
        If  the search bar has an input from the user the request is sent to the backend using fetch (post method) 
            the searchBar and mediaOption is set to state and sent to the backend to be used in the api query.
        The results from the API are then set in state in the itemsReturnedFromSearch array
    */
    handleSearchBarSubmit(event) {
        if (this.state.searchBar === "") {
            return alert("Oops! Looks like you're feeling indecisive. Please enter a search item.")
        } 
        this.setState({ itemsReturnedFromSearch: [] });
        event.preventDefault();
        fetch('http://localhost:3001/searchApi/search', {
            method: "POST",
            headers:{'Content-Type': 'application/json'},
            body: JSON.stringify({
                searchBar:this.state.searchBar,
                mediaOption:this.state.mediaOption
            })
        })
        .then(res => res.json())
        .then(results => {
            this.setState({ itemsReturnedFromSearch: results })
        })
        .catch(error => {
            alert("An error occured. Please try again later")
            this.setState(error)
        })
    };

    // The function below maps through the array that the requested items were set in state.
    displaySearchResults() {
        const items = this.state.itemsReturnedFromSearch.results;
        return (
            items.map((item) =>
                <DisplayResult
                    key={(item.collectionId + item.trackId)} // See comment below export
                    id={item.collectionId}
                    myKey={(item.artistId + item.trackId)} // myKey is passed as props (key is not passed as props)
                    link={item.trackViewUrl}
                    name={item.trackName}
                    artist={item.artistName}
                    artwork={item.artworkUrl100}
                />
            )
        );
    };

    render() {
        return (
            <div>
                <div className="navBar">
                    <Navbar bg="success" expand="lg" className="mb-4">
                        <Container>
                            <Form className="d-flex">
                                <Form.Control
                                    type="search"
                                    placeholder="Search"
                                    className="me-3"
                                    aria-label="Search"
                                    onChange={this.handleSearchBarChange}
                                />
                        
                                <Form.Select aria-label="Default select example" className='ms-2' onChange={this.handleMediaSelection} >
                                    <option value='all'>All</option>
                                    <option value='music'>Music</option>
                                    <option value='podcast'>Podcasts</option>
                                    <option value='movie'>Movies</option>
                                    <option value='tvShow'>TV Shows</option>
                                    <option value='shortFilm'>Short Film</option>
                                    <option value='audiobook'>Audiobook</option>
                                    <option value='ebook'>Ebooks</option>
                                </Form.Select>

                                <Button
                                    variant="outline-light"
                                    className='ms-2'
                                    onClick={this.handleSearchBarSubmit}>
                                    Search
                                </Button>
                            </Form>
                        </Container>
                    </Navbar>
                </div>
                <div className='searchResults'>
                    {/* The conditional operator below sets the condition that when the
                    the length of the itemsReturnedFromSearch array is empty (less than 0) then the
                    search results will not be displayed*/}
                    {this.state.itemsReturnedFromSearch.length !== 0 ? <div>{this.displaySearchResults()}</div> : ''}
                </div>
            </div>
        );
    };
};

export default Searchbar;

/* I get this error when I add key into the component: 
    Warning: DisplayResult: `key` is not a prop. Trying to access it will result in `undefined` being returned. If you need to access the 
    same value within the child component, you should pass it as a different prop.
 
    ***************** The below extract is taken from: https://reactjs.org/docs/lists-and-keys.html#keys *********************************
    Keys serve as a hint to React but they donâ€™t get passed to your components. If you need the same value in your component, pass it 
    explicitly as a prop with a different name:

    const content = posts.map((post) =>
    <Post
        key={post.id}
        id={post.id}
        title={post.title} />
    );
    With the example above, the Post component can read props.id, but not props.key.
    ****************************************************************************************************************************************
   
   I need to use the key in the other components so I used the name myKey to pass as a prop.

   If I remove the key then I get the following error: 
    Warning: Each child in a list should have a unique "key" prop.
*/