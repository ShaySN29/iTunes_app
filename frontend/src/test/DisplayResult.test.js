import React from "react";
import renderer from "react-test-renderer";
// Importing component to test render
import {DisplayResult, checkIfFavouriteItemExists } from '../components/DisplayResult';

// Tests if the component renders correctly
it('renders correctly', () => {
    const tree = renderer.create(<DisplayResult />)
    .toJSON();
    expect(tree).toMatchSnapshot();
}); 

// Test if the function to add an item to favourites in sessionStorage works correctly
it("function to add to favourites works", () => {
  let favouriteItems = [{"name": "test,", "myKey": 1111}, {"name":"test2", "myKey": 2222}];
  const exists = checkIfFavouriteItemExists(favouriteItems, 1111);
  expect(exists).toBe(true);
});