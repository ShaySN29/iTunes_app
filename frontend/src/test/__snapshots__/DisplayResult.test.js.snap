// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly 1`] = `
<div
  className="individualCard"
>
  <div
    className="displayCard mb-3 ms-2 card border-success"
    style={
      Object {
        "width": "18rem",
      }
    }
  >
    <div
      className="card-body"
    >
      <div
        className="card-title h5"
      />
      <div
        className="mb-2 text-muted card-subtitle h6"
      />
      <img
        className="card-img-top"
      />
      <a
        className="card-link"
      >
        More Details
      </a>
    </div>
    <button
      className="btn btn-outline-success"
      disabled={false}
      onClick={[Function]}
      type="button"
    >
      Add to Favourites
    </button>
  </div>
</div>
`;
